File: go/stack.go
package main

type Stack struct {
	list *SinglyLinkedList
}

func NewStack() *Stack {
	return &Stack{
		list: NewSinglyLinkedList(),
	}
}

func (s *Stack) Push(value string) {
	s.list.PushBack(value)
}

func (s *Stack) Pop() (string, error) {
	return s.list.PopBack()
}

func (s *Stack) Top() (string, error) {
	return s.list.Get(s.list.Size() - 1)
}

func (s *Stack) Size() int {
	return s.list.Size()
}

func (s *Stack) IsEmpty() bool {
	return s.list.IsEmpty()
}

func (s *Stack) Clear() {
	s.list.Clear()
}

func (s *Stack) Print() {
	s.list.Print()
}

func (s *Stack) Serialize() ([]byte, error) {
	return s.list.Serialize()
}

func (s *Stack) Deserialize(data []byte) error {
	return s.list.Deserialize(data)
}

// SerializeBinary сериализует стек в бинарный формат
func (s *Stack) SerializeBinary() ([]byte, error) {
	return s.list.SerializeBinary()
}

// DeserializeBinary десериализует стек из бинарного формата
func (s *Stack) DeserializeBinary(data []byte) error {
	return s.list.DeserializeBinary(data)
}

File: c++/stack.cpp
#include "stack.h"
#include <stdexcept>

Stack::Stack() : list() {}

void Stack::push(const std::string& value) {
    list.push_back(value);
}

std::string Stack::pop() {
    if (empty()) {
        throw std::runtime_error("Stack is empty");
    }
    std::string value = list.get(list.size() - 1);
    list.pop_back();
    return value;
}

std::string Stack::top() const {
    if (empty()) {
        throw std::runtime_error("Stack is empty");
    }
    return list.get(list.size() - 1);
}

bool Stack::empty() const {
    return list.empty();
}

size_t Stack::size() const {
    return list.size();
}

void Stack::clear() {
    list.clear();
}

void Stack::print() const {
    if (empty()) {
        std::cout << "Stack: [empty]" << std::endl;
        return;
    }
    
    std::cout << "Stack: [";
    for (size_t i = 0; i < list.size(); ++i) {
        std::cout << "\"" << list.get(i) << "\"";
        if (i < list.size() - 1) std::cout << " | ";
    }
    std::cout << "]" << std::endl;
}

void Stack::serialize(std::ostream& os) const {
    list.serialize(os);
}

void Stack::deserialize(std::istream& is) {
    list.deserialize(is);
}

void Stack::serializeText(std::ostream& os) const {
    list.serializeText(os);
}

void Stack::deserializeText(std::istream& is) {
    list.deserializeText(is);
}
